infra/experimental/sanitizers/ExecSan/execSan.cpp

```diff
@@ -17,49 +17,117 @@
 
 /* C standard library */
 #include <errno.h>
+#include <signal.h>
 #include <stdio.h>
 #include <string.h>
-#include <signal.h>
 
 /* POSIX */
-#include <unistd.h>
 #include <sys/stat.h>
 #include <sys/user.h>
 #include <sys/wait.h>
+#include <unistd.h>
 
 /* Linux */
-#include <syscall.h>
 #include <sys/ptrace.h>
+#include <syscall.h>
 
 #include <fstream>
-#include <string>
+#include <iostream>
 #include <map>
+#include <set>
+#include <sstream>
+#include <string>
 #include <vector>
 
 #define DEBUG_LOGS 0
 
 #if DEBUG_LOGS
-#define debug_log(...) \
-  do { \
-    fprintf(stderr, __VA_ARGS__); fflush(stdout); \
-    fputc('\n', stderr); \
+#define debug_log(...)            \
+  do {                            \
+    fprintf(stderr, __VA_ARGS__); \
+    fflush(stdout);               \
+    fputc('\n', stderr);          \
   } while (0)
 #else
 #define debug_log(...)
 #endif
 
-#define fatal_log(...) \
-  do { \
+#define fatal_log(...)            \
+  do {                            \
     fprintf(stderr, __VA_ARGS__); \
-    fputc('\n', stderr); \
-    exit(EXIT_FAILURE); \
+    fputc('\n', stderr);          \
+    exit(EXIT_FAILURE);           \
   } while (0)
 
 // The magic string that we'll use to detect full control over the command
 // executed.
 const std::string kTripWire = "/tmp/tripwire";
+// Shell injection bug confirmed with /tmp/tripwire.
+const std::string kInjectionError = "Shell injection";
+// Shell corruption bug detected based on syntax error.
+const std::string kCorruptionError = "Shell corruption";
+
 // The PID of the root process we're fuzzing.
 pid_t g_root_pid;
+// Assuming the longest pathname is "/bin/bash".
+constexpr int kShellPathnameLength = 10;
+// Assuming the syntax error pattern is
+// within the first 100 chars of the write buffer.
+constexpr int kErrorMessageLength = 100;
+
+// Syntax error messages of each shell.
+const std::map<std::string, std::set<std::string>> kShellSyntaxErrors = {
+    {"sh",
+     {
+         ": command not found",  // General
+         ": syntax error",       // Unfinished " or ' or ` or if, leading | or ;
+         ": missing"             // Unfinished [
+         ": event not found",    // ! leads large numbers
+         ": No such file",       // Leading < or /
+     }},
+    {"bash",
+     {
+         ": command not found",  // General
+         ": syntax error",       // Unfinished " or ' or ` or if, leading | or ;
+         ": missing"             // Unfinished [
+         ": event not found",    // ! leads large numbers
+         ": No such file",       // Leading < or /
+     }},
+    {"csh",
+     {
+         ": Command not found",       // General
+         "Unmatched",                 // Unfinished " or ' or `, leading ;
+         ": Missing",                 // Unfinished {
+         "Too many",                  // Unfinished (
+         "No match",                  // Leading [
```

Comment:
"Unmatched", "Too many", and "No match" seem a bit generic. Are there more things we can match on here?

Reply:
Yes for `{` and `(`, given they are respectively special cases.
But `[`, `"`, `'`, `` ` `` can be tricky, given the full messages are:
```
% {sdfsdf
Missing }.
% (
Too many ('s.
% [
[: No match.
% "
Unmatched ".
% '
Unmatched '.
% `
Unmatched `.
%
```

Reply:
OK can just add a TODO here to make this more specific. csh seems pretty uncommon anyway so probably not a big deal.